### PR TITLE
fix: typeId error when initializing interface with literal

### PR DESCRIPTION
**Pull request by Shanks0224**

### CODE REVIEW

**Comment by Shanks0224** [comment]
@xujuntwt95329 PTAL

File: src/semantics/expression_builder.ts [hunk 1]
```diff
@@ -1270,6 +1272,9 @@ export function newCastValue(
                     );
                 }
             }
+            // iff r-value is empty.
+            // when the 'meta' of r-value is modified, the 'typeId' of r-value also needs to be modified.
+            if (isRValEmpty) from_value.type.typeId = type.typeId;
```
Comment (xujuntwt95329) [inline src/semantics/expression_builder.ts:1277]: You only change typeid if the RVal is empty, but in L1239, you may change the meta as long as the member length is not equal, so there may still be circumstances where the assigned typeid is not matched with the actual meta

File: tools/validate/wamr/validation.json [hunk 2]
```diff
@@ -2642,6 +2642,16 @@
             }
         ]
     },
+    {
+        "module": "infc_assign_obj",
```
Comment (xujuntwt95329) [inline tools/validate/wamr/validation.json:2646]: There should already be a module named "infc_assign_obj", had better add the entry to that already existed module

File: src/semantics/expression_builder.ts (hunk 1 above)
Comment (Shanks0224) [reply src/semantics/expression_builder.ts]: done

File: tools/validate/wamr/validation.json (hunk 2 above)
Comment (Shanks0224) [reply tools/validate/wamr/validation.json]: done

File: src/semantics/expression_builder.ts [hunk 3]
```diff
@@ -1270,6 +1281,8 @@ export function newCastValue(
                     );
                 }
             }
+            // when the 'meta' of r-value is modified, the 'typeId' of r-value also needs to be modified.
+            from_value.type.typeId = type.typeId;
```
Comment (xujuntwt95329) [inline src/semantics/expression_builder.ts:1285]: You put this line in the for-loop scope, this means you can ensure all object literal to become the same shape with the assigned interface type, but seems this is not always guaranteed. e.g.:
``` TypeScript
interface I {
    x?: number
    y?: boolean
    z?: stirng
}

let obj : I = {
    y: true,
    t: 'hello',
}
```
You can add the missing `x` and `z` field, and you can adjust `y`'s type, but you can't remove the field `t`, in this case the type id is not the same with the interface `I`. So I think you need more strict rule to decide when to change the typeid.

File: src/semantics/expression_builder.ts [hunk 4]
```diff
@@ -1270,6 +1281,8 @@ export function newCastValue(
                     );
                 }
             }
+            // when the 'meta' of r-value is modified, the 'typeId' of r-value also needs to be modified.
```
Comment (xujuntwt95329) [inline src/semantics/expression_builder.ts:1284]: Another thing is L1268, seems we always push `undefined` to the end of the object literal, but the actual missing field may be in the middle of the interface, e.g.:

``` TypeScript
interface I {
    x?: number
    y?: boolean
    z?: stirng
}

let obj: I = {
    x: 1,
    z: 'h',
}
```

Adding 'z' to the end of the object literal can't ensure its typeid to be equal with interface `I`

**Review by xujuntwt95329 — APPROVED** [review]
LGTM

**Pull request merged**